### PR TITLE
MR24BHP1 Human-Presence Sensor

### PR DESCRIPTION
**Pull request by lorki97**
# What does this implement/fix?

This PR adds support for the seeedstudio MR24HPB1 Human-Presence sensor.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2322

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
# Example config.yaml
    # UART is required for MR24HPB1
    uart:
      tx_pin: 17
      rx_pin: 16
      baud_rate: 9600

    # Example configuration entry
    mr24hpb1:
      threshold_gear: 7
      scene_setting: DEFAULT
      force_unoccupied: NONE
      environment_status:
        name: Environment Status
      occupancy:
        name: Occupancy
      movement:
        name: Movement
      movement_rate:
        name: Movement Rate
      movement_type:
        name: Movement Type
      device_id:
        name: Device ID
      software_version:
        name: Software version
      protocol_version: 
        name: Protocol Version
      hardware_version: 
        name: Hardware Version
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by andyfsimon** [comment]
This is precisely what I was looking for - I'd love to test it, but I am not practical with esphome-dev.

**Comment by lorki97** [comment]
@andyfsimon You would have to clone the code from https://github.com/lorki97/esphome/tree/feature/Seeed-Studio-MR24D11C10 and setup a development environment on your PC to use this module.

You can find the configuration documentation here: https://deploy-preview-2322--esphome.netlify.app/components/mr24hpb1.html (from https://github.com/esphome/esphome-docs/pull/2322)

**Comment by singhdd93** [comment]
Any Idea when this will be merged ?

**Comment by andyfsimon** [comment]
I managed to build the code and it does indeed work. I still have to figure out the parameters though, as it seems to be detecting pretty much anything and not only presence. For instance, while pointed to an empty room with a 3d printer running it'll go off. Might be legit, but that's not what I expected.
Other than that, it seems to be working just great so thumbs up! I'll keep testing.

**Comment by andyfsimon** [comment]
Indeed I do see a potential issue, but I am unsure of what's causing this: no matter what threshold level I select the minimum detection of movement rate is 1. Meaning that even if I leave the room (empty) it will still detect presence, albeit stationary.

**Comment by florianL21** [comment]
I've found that the sensor is quite sensitive even in the lowest sensitivity level.
I currently have it mounted at a 15 degree angle right below the ceiling which in my room I've found works best.
I also have the same "issue" that it will never switch to unoccupied when my 3d printer is running but I guess that is to be expected since it's a bunch of metal moving around and the sensor is trying to detect micro movements.

I've also found that the sensor seems to be quite picky about where it is placed. It might not work in some locations/orientations at all. I suggest trying different locations if you are having the issue that the room never goes to unoccupied.

The sensor also does not like to be moved/vibrated after it's started. So make sure it is mounted to a steady location. If you readjust it's position try to cut power for a few seconds and turn it back on again. There seems to be a background noise analysis ongoing in the first 30s after the sensor starts. So make sure to also keep it really steady as soon as it has power.

Hope these tips will help you to get it working reliably

**Comment by andyfsimon** [comment]
Hi @florianL21, you make some very interesting points indeed. I did make sure to create a generous separation (for tests) between the ESP32 and the sensor (I've been told the ESP antenna interferes with the sensor at times) but I did not consider vibrations or, more interestingly, the need to keep it steady during startup.
I tried removing the printer, which was the most obvious interference, but I suppose other elements in the room or outside could be making the sensor's life harder than it should.
I was hoping to be able to test before putting it into the enclosure I printed (which contains other sensors as well) but it seems I'll have to assemble and place the unit before I can test/calibrate.

Thanks a lot for the suggestions and of course also to @lorki97 for the great work! I hope it'll be merged soon!

File: esphome/components/mr24hpb1/constants.h [hunk 1]
```diff
@@ -0,0 +1,210 @@
+#pragma once
+
+#include "mr24hpb1.h"
```
Comment (jesserockz) [inline esphome/components/mr24hpb1/constants.h]: Why include the whole component when only "types.h" is needed here?

**Comment by asucrews** [comment]
Really lookin forward to this integration! I hope @lorki97 has the time to review Jesse comments.

**Comment by lorki97** [comment]
> Really lookin forward to this integration! I hope @lorki97 has the time to review Jesse comments.

Thanks for pinging me, the notifications got drowned in my email folder. I'll check this ASAP.

File: esphome/components/mr24hpb1/constants.h (hunk 1 above)
Comment (florianL21) [reply esphome/components/mr24hpb1/constants.h]: @lorki97 updated this already, I just checked it again on the actual hardware and everything still seems to be working fine

**Comment by asucrews** [comment]
Bump

File: esphome/components/mr24hpb1/__init__.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+            cv.Optional(CONF_ENVIRONMENT_STATUS): text_sensor.text_sensor_schema(),
+            cv.Optional(CONF_OCCUPANCY): binary_sensor.binary_sensor_schema(
+                device_class=DEVICE_CLASS_OCCUPANCY,
+            ),
+            cv.Optional(CONF_MOVEMENT): binary_sensor.binary_sensor_schema(
+                device_class=DEVICE_CLASS_MOTION,
+                icon=ICON_MOTION_SENSOR,
+            ),
+            cv.Optional(CONF_MOVEMENT_RATE): sensor.sensor_schema(),
+            cv.Optional(CONF_MOVEMENT_TYPE): text_sensor.text_sensor_schema(),
```
Comment (jesserockz) [inline esphome/components/mr24hpb1/__init__.py]: Please split these entities out into their own platform files similar to how other components do inside ESPHome.

File: esphome/components/mr24hpb1/__init__.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    DEVICE_CLASS_OCCUPANCY,
+    DEVICE_CLASS_MOTION,
+    ICON_MOTION_SENSOR,
+    ENTITY_CATEGORY_DIAGNOSTIC,
+)
+
+CODEOWNERS = ["@lorki97", "@florianL21"]
+
+DEPENDENCIES = ["uart"]
+AUTO_LOAD = ["text_sensor", "binary_sensor", "sensor"]
```
Comment (jesserockz) [inline esphome/components/mr24hpb1/__init__.py]: `AUTO_LOAD`ing multiple platforms like this is not allowed in new components unless absolutely necessary, which is almost never if the code is written right.
```suggestion
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/mr24hpb1/__init__.py (hunk 3 above)
Comment (lorki97) [reply esphome/components/mr24hpb1/__init__.py]: Removed in newest commit.

File: esphome/components/mr24hpb1/__init__.py (hunk 2 above)
Comment (lorki97) [reply esphome/components/mr24hpb1/__init__.py]: Split up into own platform files according how other components handle this. Some testing is required to ensure this didn't change anything else. I currently don't have access to the required HW, maybe @florianL21 or somebody else can test with the newest commit.

**Comment by asucrews** [comment]
@lorki97 did this ever get merged?

**Comment by lorki97** [comment]
@asucrews This has not been merged. I have little hope of getting this merged anytime soon with the PR being opened nearly 1.5 years ago.

**Pull request closed without merging**